### DISCUSSION TITLE
Python: universal wheels are better than building from source

### DISCUSSION THREAD

**Discussion by adamjstewart** [spack/spack#30795] [General]
Spack has always primarily been a "from source" package manager. Up until now, the advice for Python libraries was to build from source whenever possible and only use wheels when the source code is unavailable. However, it may be time to stop treating Python like a compiled programming language and use wheels instead.

### Pros

There are many advantages of installing wheels directly:

1. **Avoid circular dependencies.** We already install [setuptools/wheel/pip](https://github.com/spack/spack/pull/27798) from wheels for this very reason.
2. **Avoid conflicting versions of build dependencies.** Many build systems like jupyter-packaging encourage developers to pin a specific version of the build system in `pyproject.toml`. Since Spack can't yet concretize multiple versions of the same package in a DAG, I was forced to split [jupyter-packaging](https://github.com/spack/spack/pull/27798) into multiple packages. Same story for [tensorboard-plugin-wit](https://github.com/spack/spack/pull/28566) which requires multiple versions of bazel in the DAG.
3. **Avoid non-Python dependencies.** In order to build tensorboard, I need to build bazel (which requires java to build) and tensorboard-data-server (which requires rust to build). These language compilers are massive, take forever to build, and you may find that [tensorboard-data-server](https://github.com/spack/spack/pull/29596) requires a different version of Rust than you have. I've encountered similar issues with [jupyterlab-widgets](https://github.com/jupyter-widgets/ipywidgets/issues/3324) and many other packages in the Jupyter ecosystem which rely on npm to build many of their wheels, although the wheels themselves are OS-independent.
4. **Faster build times.** The aforementioned Rust and Bazel can take several hours to build from source, but aren't needed if we just install from wheels. More concretely, I just build tensorboard from source and it took 2.4 hours, but the installed library is pure-Python and could be installed from a wheel in under 1 second with no difference in the installed files. Why?
5. **Fewer issues on esoteric architectures.** [bazel](https://github.com/spack/spack/pull/28548#issuecomment-1130491322), [tensorboard](https://github.com/tensorflow/tensorboard/issues/5714), and [tensorboard-data-server](https://github.com/tensorflow/tensorboard/issues/5713) don't build on M1. But tensorboard and tensorboard-data-server are both pure-Python libraries with universal wheels, so there's no reason to try to "build" them at all.
6. **Automatically find dependencies.** For packages that use `setup.cfg` or `pyproject.toml` to list dependencies, we can parse these files to determine the package dependencies, but for packages that use `setup.py`, it's impossible to determine these dependencies without running `setup.py`. But if we install from wheels instead, these dependencies are already listed in the wheel for us. This would allow us to autogenerate all Python libraries with `spack create`. We could also check all packages for missing dependencies in CI. See [here](https://discuss.python.org/t/list-dependencies-of-a-package/12341) for more info.
7. **URLs get simpler.** Some packages don't always have source code available for every version so we had to download from GitHub sometimes, or switched from `.zip` files to `.tar.gz` files so we had to add a `url_for_version`. If we only need `.whl` files this will be simpler.

### Cons

* **No support for `spack develop`.** You can't iteratively make changes to a package and rebuild it if it doesn't use the source code. However, `spack develop` isn't particularly useful for pure-Python libraries. As long as these packages are in your current directory or somewhere in your `PYTHONPATH`, they are already installed, there is no compilation necessary.

### Exceptions

At this point, I should mention the obvious caveat that not all Python libraries are pure-Python. For packages like [numpy](https://pypi.org/project/numpy/#files), [scipy](https://pypi.org/project/scipy/#files), [PyTorch](https://pypi.org/project/torch/#files), and [TensorFlow](https://pypi.org/project/tensorflow/#files), we unfortunately don't have a choice. We could use the arch/platform-specific wheels available for these, but those won't work on esoteric architectures (many packages don't yet publish M1 wheels), and we would suffer from poor performance due to non-optimized binaries.

My thinking is that for packages that offer a single universal wheel (`*-py3-none-any.whl`), we should install from that instead of from source code. For packages that offer a wheel for each arch/platform, we should continue to install from source. Where this gets interesting is packages like [cython](https://pypi.org/project/Cython/#files) or [tensorboard-data-server](https://pypi.org/project/tensorboard-data-server/#files) which offer both universal wheels and arch/platform-specific wheels. I'm not sure what to do with those kinds of packages.

@tgamblin has complained about wheels before and may have opinions.
@rgommers may have more knowledge of cython/wheels than I do.
@glennpj @aweits this conversation directly pertains to many of the packages you maintain (which I've had a hell of a time installing on M1). Is there any reason we shouldn't just install these from wheels?

**Comment by rgommers** [comment]
> My thinking is that for packages that offer a single universal wheel (`*-py3-none-any.whl`), we should install from that instead of from source code. For packages that offer a wheel for each arch/platform, we should continue to install from source. 

Agreed, this is clearly a good idea.

> Where this gets interesting is packages like [cython](https://pypi.org/project/Cython/#files) or [tensorboard-data-server](https://pypi.org/project/tensorboard-data-server/#files) which offer both universal wheels and arch/platform-specific wheels. I'm not sure what to do with those kinds of packages.

Cython can work as pure Python, but is faster when compiled (it compiles the pure Python code with Cython itself). Not sure about tensorboard. It would be sensible to assume that if there are non-universal wheels on PyPI, those offer extra performance or functionality, and therefore it's better for Spack to build from sdist in those cases.

**Comment by alalazo** [comment]
Relates to this old issue https://github.com/spack/spack/issues/1249

**Comment by tgamblin** [comment]
I'm strongly opposed to this, but I'm strongly for *binaries*.  Our own binaries.

Wheels are basically another package manager's unmanaged, unsecured, unvetted binary distribution.  Well, not even that -- they're whatever the PyPI maintainers put on PyPI.  Yes, they're quick and dirty, and they install fast.  No, they don't fit the Spack model -- which requires reproducibility, provenance, and parameterizable, portable builds.  There's a nice dissection of what's in wheels (or at least the PyTorch one) in [this article](https://hpc.guix.info/blog/2021/09/whats-in-a-package/) by the Guix folks.

If we go with wheels, we're going to lose all the ability to port and integrate python packages with HPC codes on new platforms, mainly because wheels are not designed for those platforms.

On @adamjstewart's list:

> 1. **Avoid circular dependencies.** We already install [setuptools/wheel/pip](https://github.com/spack/spack/pull/27798) from wheels for this very reason.

Yeah, we do, and we really shouldn't long-term.  What we need to be able to do, eventually is self-host this stuff from *spack* binaries.  Right now we cannot do that, so we do workarounds like this.  But *very soon* we are going to have a public binary cache, and I think the way to do this is to build a Spack binary (possibly using external packages, but with provenance) and to use that as the bootstrap case instead of wheels.

> 2. **Avoid conflicting versions of build dependencies.** [...] Since Spack can't yet concretize multiple versions of the same package in a DAG, I was forced to split [jupyter-packaging](https://github.com/spack/spack/pull/27798) into multiple packages. [...]

True.  *but*, the infrastructure for this is in place as of #22845.  We originally keyed deps in a DAG by name, but now they are keyed by hash, which essentially lets us have arbitrarily many.  We need to work that into the concretizer -- #28941 is a start to that, and we expect to have separate build deps by 0.19, as we need it for compilers as deps.

> 3. **Avoid non-Python dependencies.** 

For portability... we don't *want* to avoid non-Python dependencies.  If we do, we're tying ourselves to whatever wheels are available, and we're not going to be able to port these stacks to new systems easily.  We'll basically have to build wheels.  We do not want to tie the Spack supply chain to PyPI or to wheels -- we want it traceable to source.

> 4. **Faster build times.** Why?

Because we're not building?  The way to avoid this is to build binaries.  We're going to have a public binary cache very soon here (weeks), and I think the way to get Bazel to install fast is to, well, build it in the build farm. Then we can tweak the build farm to build the variants we need for things like Bazel, and also we can just install tensor board from a binary without bothering with the build deps.

> 5. **Fewer issues on esoteric architectures.**

Sure, for the pure python packages.  But for everything else (which I would argue is why people use Spack), using wheels is going to give you *more* issues on esoteric architectures, because you'll lose the ability to build things.  We *really* don't want to do that -- Spack isn't a binary package manager; it's a source package manager with binary caches. We *want* users to be able to seamlessly mix sources and binaries and if we take prepackaged binaries we know nothing about, we'll lose that (as you say above).

> 6. **Automatically find dependencies.** 

This is the most fundamental thing I would like to automate, but I don't (yet) know how. Python packages basically don't have parseable metadata, as you note.  Particularly with `setup.py`, it can be very hard to figure out what the dependencies of a package are, whether some are optional, etc.  Wheels solve this, but only for a particular configuration of the package -- you won't get conditionals or any of the stuff we put in Spack packages out of them.  *Maybe* we could parse wheel metadata for a lot of wheels, and convert that to Spack directives.  Would be interested to hear about automated ways to do this.

Most of these are issues with packages that aren't pure python, but there are features in Spack that we want to enable that wheels directly conflict with.  The main one is `spack develop`.  Pure python wheels don't necessarily look like the source tree for the package, so you can't `spack develop <pkg>` and get a source checkout that looks like anything you'd submit a pull request from.  This is my main reason for opposing pure python wheels -- they're not actually the package source.  They're a downstream distribution.  That said, I would probably compromise on this one if we had some way to avoid maintaining pure python packages altogether.


So my suggestion here is that we should:
1. Wait a week or two until the public build cache is deployed, where we'll be offering binaries from our own build farm.  That alone is already a leg up on PyPI, which has no build farm and takes whatever binaries people upload.
2. Get all the python packages and their (potentially non-python) deps in the buildcache.  I think this will make installing python stuff with spack a lot better, *and* it'll preserve our ability to customize packages with variants, etc.
3. We're going to have separate build deps in 0.19, which I think will ease the pains of splitting packages like `jupyterhub`.
4. I would *still* really like to find a way to automate finding metadata for pure python packages -- e.g., I *think* we could do them as a custom Spack `Repository` class, and not have to maintain Python packages much at all.  This is a bigger, longer-term thing, but I want something that would let us leverage PyPI metadata and override some default python-only build for packages.

**Reply by adamjstewart** [reply]
@tgamblin I think the vast majority of your concerns can be addressed by the fact that the above proposal is limited to pure-Python libraries. The following are responses to individual comments:

1. **Avoid circular dependencies:** it's unclear to me how self-hosted binaries will help since Spack still needs to be able to build those from source where we have circular deps.
2. **Avoid conflicting versions of build dependencies:** really excited to see a concretizer that allows build deps in the DAG multiple times.
3. **Avoid non-Python dependencies:** in the case of tensorboard and Jupyter, these are pure-Python libraries that don't use non-Python libraries at run-time, only at build-time. They are already platform independent, hence why there are universal wheels for these libraries. For example, npm is only needed to download JavaScript files or something silly like that.
4. **Faster build times:** I agree that public binaries for packages will solve many of the concerns about build times.
5. **Fewer issues on esoteric architectures:** unfortunately public binaries won't help here. tensorboard and its bazel dependency simply don't build on M1, despite tensorboard being a pure-Python package. And Google is notoriously bad about platform support (TensorFlow is supported on Debian/Ubuntu but not RHEL/CentOS/Fedora).
6. **Automatically find dependencies:** there are some tools to help find imports in packages, but realistically there's not much we can do here without either using wheels, or building wheels ourselves in a build farm.

To reiterate, the above proposal is for pure-Python libraries with universal wheels, not for platform/OS/Python-version-dependent wheels which we will still need to build from source (that's the whole point of Spack).

My other concern with the build cache is platform support. Since we support granularity at the level of specific CPU model, it will be quite challenging to ensure that we build binaries for all versions/variants of all packages on all platforms. I'm not sure about our other resources, but GitHub Actions still doesn't support M1, so we can't even properly test bootstrapping. I guess I'm just pessimistic, but I want to make sure that Spack still works great even if there isn't a public binary cache available for your system.

**Reply by adamjstewart** [reply]
Also, regarding the concerns about `spack develop`, I don't think the `spack develop` pipeline is useful or necessary for pure-Python libraries. For these packages, as long as the package is in your current directory or in your `PYTHONPATH`, it is already installed. So the use of `spack develop` shouldn't be a reason to avoid universal wheels.

**Comment by rgommers** [comment]
There's a major discrepancy in the issue description by @adamjstewart and his actual proposal. The former is mostly about the advantages of binary wheels, while the proposal ended up being (paraphrasing) _"use wheels instead of sdists for pure Python packages, change nothing otherwise"_.

A universal wheel is (99% of the time) simply a "better sdist for pure Python packages", because static metadata, no arbitrary Python code execution at build time, etc. It's also no less trustable than the sdist itself (both uploaded by the same random person) - if you want to improve on that, the way to go is to install from the release tag in VCS rather than from PyPI. But [the PyPI vs. GitHub doc section](https://github.com/spack/spack/blob/8556f832380ce77a9ccdd97386ff7f774f956cba/lib/spack/docs/build_systems/pythonpackage.rst#pypi-vs-github) prefers PyPI.

**Reply by adamjstewart** [reply]
> if you want to improve on that, the way to go is to install from the release tag in VCS rather than from PyPI. But [the PyPI vs. GitHub doc section](https://github.com/spack/spack/blob/8556f832380ce77a9ccdd97386ff7f774f956cba/lib/spack/docs/build_systems/pythonpackage.rst#pypi-vs-github) prefers PyPI.

Yeah, I wrote that, so I'm biased. Honestly, I consider PyPI sdist and wheels to be _more_ secure than release tags on VCS since the former cannot be modified after it is released while the latter is a moving target, making it extremely difficult to checksum properly.

**Comment by glennpj** [comment]
I think what is being said is that pure python packages are not able to build on M1 because the build tools do not build on M1. Perhaps those pure python packages could be tagged and treated special for the Spack build cache. In other words, if they are pure python then they only need to be built once for the build cache and can then be installed on any platform. Instead of universal wheels they would be universal spack binaries.

**Reply by adamjstewart** [reply]
@tgamblin any thoughts on this proposal? I'm not sure how long this would take to implement, but I really do need to use tensorboard and I would rather not use pip to install it. Should I convert `py-tensorboard` to install a wheel in the meantime? There are absolutely no benefits of building it from source other than a 30+ minute compile time and failure on systems like Apple M1.

**Comment by tgamblin** [comment]
I talked to @adamjstewart and thought about this a bit more and I think I'm convinced that wheels are the right way to go.

My main concerns are:
1. **Packages with native code**: I think we all (mostly?) agree we should keep building these packages with Spack, otherwise we lose the ability to ensure that their builds are consistent with the rest of the DAG.  
2. `spack develop`: We already have to differentiate between dev builds and regular builds for `autotools` packages, where the source distribution contains a generated `configure` script, but the git checkout doesn't.  So if we have to do something similar with python packages, so that if we check out from a repo we need the build tools, that's not any worse.  I guess most packages won't have this, but they can grow dev builds as needed, and we can get stuff like `tensorboard` in Spack faster with wheels.
3. **Hashes:** We guarantee in Spack that source distributions have a local `sha256` hash so that downloads can be verified.  If we ensure that we add this for wheels, I think I'm fine with them.  `pip` notably doesn't check hashes by default -- they *just* rely on `SSL`.  I think we should use stronger validation than that, or at least use hashes to be consistent with other Spack packages.

I think I'm convinced that universal wheels are just a better-packaged source distribution (i.e., you can just unzip them), so if we avoid binary wheels and just use universal ones, and if we ensure there are checksums, then I think this is fine.  We can still make Spack binaries from wheels (and include them in build caches).

The other really compelling advantage mentioned here is that wheels have metadata, so we could vastly simplify the creation and maintenance of Python packages using that.  My worries there are still that:
1. My impression is that a lot of Python package metadata is overconstrained -- e.g. they only allow you to install with specific versions of dependencies, which is going to make integration hard.
2. We at least need some way to update a Python package with a new version with a new wheel and its metadata.  I haven't thought it through completely, but I think converting constraints and merging metadata may be nontrivial.  But it seems doable.

So go ahead @adamjstewart -- let's make it easy to use universal wheels for Python packages.  We'll still need traditional Spack packages for the built stuff, but at least this will focus the effort around the hard parts and make all the universal packages easier to maintain.

**Reply by adamjstewart** [reply]
I'm going to start a separate discussion for how to auto-create Python (etc.) package with `spack create` complete with dependencies for a range of versions. In the meantime, I'll update tensorboard to install from wheels.

**Reply by tgamblin** [reply]
For those interested, the follow-up is here: https://github.com/spack/spack/discussions/31413
